### ISSUE TITLE
How to avoid processing the favicon requests?

### ISSUE THREAD

**Issue by AlexPravdin** [spiral/roadrunner#85]
Browsers always tries to load the `favicon.ico` file from the root. How to avoid passing these requests to PHP?

**Comment by wolfy-j** [comment]
RR passes all of non static requests to the workers, it does not contain routing logic rather than provided by static service. You can do it multiple ways:
1) put real favicon.ico
2) see @grachevko response
3) implement custom middleware on Golang end

**Comment by AlexPravdin** [comment]
@grachevko Thanks for the feedback. But using this logic we don't need the `static` config section as this functionality can be made on top of RR.

As we already have the `static` configuration it would be fine to have a list of particular files in this section that RR will not handle in addition to directory configuration.

**Comment by wolfy-j** [comment]
@AlexPravdin it already has such list, see `forbid`:

```yaml
# static file serving. remove this section to disable static file serving.
static:
  # root directory for static file (http would not serve .php and .htaccess files).
  dir:   "public"

  # list of extensions for forbid for serving.
  forbid: [".php", ".htaccess"]
```

**Comment by wolfy-j** [comment]
Unless you mean extension which static service MUST always handle.

**Comment by AlexPravdin** [comment]
@wolfy-j I was tried to use the `forbid` option but it does'n worked for me. 
I have the following file:

`public/1.html`

And the following `static` section:

```
static:
  dir:   "."
  forbid: [".php", ".htaccess", ".ico"]
```

But requests to `public/1.html` are never served to PHP (at least breakpoint in the first line of the `$psr7->acceptRequest()` loop is not reached) while the `html` extension is not listed in the `forbid` parameter (and is it a list of extensions or a list of file names?). Requests to `favicon.ico` are always served to PHP while `.ico` extension is listed in the `forbid` parameter. What I'm doing wrong?

**Comment by wolfy-j** [comment]
This is inverse logic, `forbid` forbids static service to server files of specific extension *as files*, meaning requests will always be passed to PHP. We can add another directive such as `always` for the logic you need.

**Comment by AlexPravdin** [comment]
@wolfy-j Yes, an option to list files to always treat them as static even they are not exists would be very helpful for my case to avoid noisy files whose purpose is only to improve the app server functionality.

**Comment by AlexPravdin** [comment]
@wolfy-j Could you also mention please in the docs that the `forbid` option is related to the static service only? This comment:

> list of extensions for forbid for serving

is not so clear to understand that you mean serving by the static service, not by http service.

**Comment by wolfy-j** [comment]
By static service, sorry for the confusion.

**Comment by wolfy-j** [comment]
@AlexPravdin i'm going to name this option "always", but i'm not sure if this is the best name:
```yaml
always: [".ico"]
```

Any ideas?

**Comment by mnavarrocarter** [comment]
Personally, I handle this in my PHP. My first middleware is a static file middleware. Works like a charm.

**Comment by AlexPravdin** [comment]
@wolfy-j As from my point of view, the clear option set might be like this:

```
static:  # Static files options to prevent sending them to PHP
  include: ['/exact-dir-in-root', '/exact-file.html', '/all-files-in-dir/*', 'all-dirs-and-files-with-this-text', '*.ext'] # list of directory and file name masks with wildcards (relative to the config dir)
  exclude: ['/public/upload/*.php'] # masks of files/dirs explcitly not treated as static
```

File mask syntax can be taken from `.gitignore`, for example, to be familiar for most developers.

**Comment by wolfy-j** [comment]
File masks would be outside of the scope of what i have already build. For now i will keep default implementation with ability to replace static server with more sophisticated implementation.

**Comment by wolfy-j** [comment]
Fixed in 1.3.1, use `always` directive of static service as mentioned above.

**Comment by AlexPravdin** [comment]
Problem solved, thanks!
